### DISCUSSION TITLE
Can I use environment variable for server IP?

### DISCUSSION THREAD

**Discussion by albertski** [basecamp/kamal#509] [Q&A]
Is it possible to use environment variables for the server IPs? For example,
```
# Deploy to these servers.
servers:
  web:
    - KAMAL_DEPLOY_IP_1
```

I get the following error now if I deploy when using the environment variable:
```
  ERROR (SocketError): Exception while executing on host KAMAL_DEPLOY_IP_1: getaddrinfo: nodename nor servname provided, or not known
```

**Comment by medius** [comment]
You can use ERB in deploy.yml to dynamically insert anything including server IPs

```ruby
# Deploy to these servers.
servers:
  web:
    - <%= ENV['KAMAL_DEPLOY_IP_1'] %>
```

Edited: As mentioned below, I had accidentally used the wrong syntax earlier and have now corrected it.

**Reply by cosmin-marginean** [reply]
This doesn't seem to work, maybe because of `image` not supporting this?
```
image:
  - <=% ENV['DOCKER_IMAGE'] %>
```
(Sorry, not a Ruby expert, not sure of this is correct)

**Reply by ttilberg** [reply]
Note that OP had the tag syntax switched. It should be `<%=` as opposed to `<=%`.

**Comment by crohr** [comment]
Yes, definitely doable, even with multiple IPs, for instance:

```
servers:
  web:
    hosts: <%= ENV.fetch('SERVER_IPS').split(' ') %>
```

You could go as far as to fetch those IPs from your hosting provider API, although I wouldn't necessarily recommend it.

**Comment by hipertracker** [comment]
OK, fine but how to pass env variables from Kamala secrets and 1Password? I can define env vars for the whole service but what but not only for the specific server. I define it currently globally in deploy.yml (this is for Django app)

```
env:
  clear:
    DEBUG: 0
  secret:
    - DATABASE_URL
```

**Reply by emerson-argueta** [reply]
Looking at the source code, this wouldn't work, but we can create an issue and propose to pull from the secrets config.
For reference I'm looking at the servers config:
https://github.com/basecamp/kamal/blob/main/lib/kamal/configuration/servers.rb

and we can propose to do similarly what is done for registry config:
https://github.com/basecamp/kamal/blob/main/lib/kamal/configuration/registry.rb
